contrib/vcl/source/vcl_interface.cc

```diff
@@ -0,0 +1,122 @@
+namespace Network {
+namespace Vcl {
+
+namespace {
+
+/**
+ * Max number of workers supported. Used in declaration of epoll_handles array
+ */
+const int MaxNumWorkers = 128;
+
+/**
+ * Max number of epoll events to drain from VCL per `vppcom_epoll_wait` call
+ */
+const int MaxNumEpollEvents = 128;
+
+/**
+ * Envoy worker epoll session handles by VCL worker index, i.e., `vppcom_worker_index()`. Each
+ * worker uses its respective handle to retrieve session events from VCL via `vppcom_epoll_wait()`.
+ */
+uint32_t epoll_handles[MaxNumWorkers];
+
+/**
+ * Mutex only used during VCL worker registration
+ */
+ABSL_CONST_INIT absl::Mutex wrk_lock(absl::kConstInit);
+
+using MqFileEventsMap = absl::flat_hash_map<int, Envoy::Event::FileEventPtr>;
+
+MqFileEventsMap& mqFileEventsMap() { MUTABLE_CONSTRUCT_ON_FIRST_USE(MqFileEventsMap); }
+
+void onMqSocketEvents(uint32_t flags) {
+  ASSERT((flags & (Event::FileReadyType::Read | Event::FileReadyType::Write)));
+  auto wrk_index = vppcom_worker_index();
+  VCL_LOG("events on worker {}", wrk_index);
+  struct epoll_event events[MaxNumEpollEvents];
+  int max_events = MaxNumEpollEvents, n_events;
+
+  while (max_events > 0) {
+    n_events = vppcom_epoll_wait(epoll_handles[wrk_index], events, max_events, 0);
+    if (n_events <= 0) {
+      break;
+    }
+    max_events -= n_events;
+    VCL_LOG("had {} events", n_events);
+
+    for (int i = 0; i < n_events; i++) {
+      VclIoHandle* vcl_handle = reinterpret_cast<VclIoHandle*>(events[i].data.u64);
+      if (vcl_handle->isWrkListener()) {
+        vcl_handle = vcl_handle->getParentListener();
+      }
+
+      // session closed due to some recently processed event
+      if (!vcl_handle->isOpen()) {
+        continue;
+      }
+
+      uint32_t evts = 0;
+      if (events[i].events & EPOLLIN) {
+        evts |= Event::FileReadyType::Read;
+      }
+      if (events[i].events & EPOLLOUT) {
+        evts |= Event::FileReadyType::Write;
+      }
+      if (events[i].events & (EPOLLERR | EPOLLHUP)) {
+        evts |= Event::FileReadyType::Closed;
+      }
+
+      VCL_LOG("got event on vcl handle fd {} sh {:x} events {}", vcl_handle->fdDoNotUse(),
+              vcl_handle->sh(), evts);
+      vcl_handle->cb(evts);
+    }
+  }
+}
+
+} // namespace
+
+uint32_t vclEpollHandle(uint32_t wrk_index) { return epoll_handles[wrk_index]; }
+
+void vclInterfaceWorkerRegister() {
+  {
+    absl::MutexLock lk(&wrk_lock);
+    vppcom_worker_register();
+  }
+  int epoll_handle = vppcom_epoll_create();
+  epoll_handles[vppcom_worker_index()] = epoll_handle;
+  VCL_LOG("registered worker {} and epoll handle {:x} mq fd {}", vppcom_worker_index(),
+          epoll_handle, vppcom_mq_epoll_fd());
+}
+
+void vclInterfaceRegisterEpollEvent(Envoy::Event::Dispatcher& dispatcher) {
+  MqFileEventsMap& mq_fevts_map = mqFileEventsMap();
+  if (mq_fevts_map.find(vppcom_worker_index()) != mq_fevts_map.end()) {
+    return;
+  }
+  RELEASE_ASSERT(vppcom_worker_index() != -1, "");
```

Comment:
Better do asserting before the first use of the index. Also we could have a const initalized once with `vppcom_worker_index()` and then use the const.